presto-main/src/main/java/com/facebook/presto/metadata/ThriftRemoteNodeState.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.drift.client.DriftClient;
+import com.facebook.presto.server.thrift.ThriftServerInfoClient;
+import com.facebook.presto.spi.NodeState;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.Duration;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.units.Duration.nanosSince;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class ThriftRemoteNodeState
+        implements RemoteNodeState
+{
+    private final ThriftServerInfoClient thriftClient;
+    private final AtomicReference<Optional<NodeState>> nodeState = new AtomicReference<>(Optional.empty());
+    private final AtomicReference<Future<?>> future = new AtomicReference<>();
+    private final AtomicLong lastUpdateNanos = new AtomicLong();
+
+    public ThriftRemoteNodeState(DriftClient<ThriftServerInfoClient> thriftClient, URI stateInfoUri)
+    {
+        requireNonNull(stateInfoUri, "stateInfoUri is null");
+        checkState(stateInfoUri.getScheme().equals("thrift"), "unexpected scheme %s", stateInfoUri.getScheme());
+
+        this.thriftClient = requireNonNull(thriftClient, "thriftClient is null").get(Optional.of(stateInfoUri.getAuthority()));
+    }
+
+    @Override
+    public Optional<NodeState> getNodeState()
+    {
+        return nodeState.get();
+    }
+
+    @Override
+    public synchronized void asyncRefresh()
+    {
+        Duration sinceUpdate = nanosSince(lastUpdateNanos.get());
+
+        if (sinceUpdate.toMillis() > 1_000 && future.get() == null) {
```

Comment:
Correct me if I'm wrong, but we're using the `AtomicReference` just to check if something's in flight, i.e. we don't actually use that future.  If that's the case, why not just use an `AtomicBoolean`?  You could also remove the thread-level synchronization as well if you just atomically set false->true in this if statement.